src/MultiphaseStaggeredHierarchyIntegrator.cpp

```diff
@@ -875,19 +903,55 @@ MultiphaseStaggeredHierarchyIntegrator::preprocessIntegrateHierarchy(const doubl
         const int xi_idx = var_db->mapVariableAndContextToIndex(d_xi_var, getScratchContext());
         d_xi_fcn->setDataOnPatchHierarchy(xi_idx, d_xi_var, d_hierarchy, eval_time, false, coarsest_ln, finest_ln);
     }
-
-    MultiphaseStaggeredStokesOperator RHS_op("RHS_op", true, d_params);
+    // Boundary Condition helper.
     Pointer<StaggeredStokesPhysicalBoundaryHelper> bc_helper = new StaggeredStokesPhysicalBoundaryHelper();
-    RHS_op.setPhysicalBoundaryHelper(bc_helper);
-    RHS_op.setPhysicalBcCoefs(d_un_bc_coefs, d_us_bc_coefs, nullptr, d_thn_bc_coef);
-    // Divergence free condition and pressure are not time stepped. We do not need to account for the contributions in
-    // the RHS.
-    RHS_op.setCandDCoefficients(C, D1, 0.0, 0.0);
-    RHS_op.setThnIdx(thn_cur_idx); // Values at time t_n
 
-    // Store results of applying stokes operator in rhs_vec
-    RHS_op.initializeOperatorState(*d_sol_vec, *d_rhs_vec);
-    RHS_op.apply(*d_sol_vec, *d_rhs_vec);
+    if (getIntegratorStep() != 0 && d_viscous_ts_type == TimeSteppingType::BDF2)
+    {
+        // If the integrator step is 0 (so initial time step), we reduce to backward Euler, and this block is skipped.
+        // Note that the old velocities are already multiplied by volume fraction.
+        const int un_old_idx = var_db->mapVariableAndContextToIndex(d_un_old_var, getCurrentContext());
+        const int us_old_idx = var_db->mapVariableAndContextToIndex(d_us_old_var, getCurrentContext());
+
+        const int rhs_un_idx = d_rhs_vec->getComponentDescriptorIndex(0);
+        const int rhs_us_idx = d_rhs_vec->getComponentDescriptorIndex(1);
+
+        double alpha = d_dt_previous[0] / dt;
```

Comment:
I suppose `d_dt_previous` is some sort of vector containing prior time step sizes (so BDF2 can be used) and is a member variable of the hierarchy integrator class in IBAMR.